[Cheah Tze Juen] iP

cheahTJ opened the pull request:
## Pingmebot

Pingmebot allows to have a peace of mind and remember the things you have to do. It's,

- text-based 
- easy to learn

All you need to do is,

1. Find the source code [here](https://github.com/cheahTJ/ip)
2. Download it
3. Start using it
4. Let it manage all your tasks 😉

And most importantly, it is **FREE**

The features include:

- [ ] Managing todo 
- [ ] Managing deadlines
- [ ] Managing events

f you are a Java programmer, you can use it to practice Java too. Here's the `main` method

```
public class Pingme {
  public static void main(String[] args) {
          new PingMe("./data/dukeData.txt").run();
   }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

cheahTJ committed:
Renamed bot, added greeting and exit msgs

cheahTJ committed:
Added echoing feature

cheahTJ committed:
Added storage of tasks

cheahTJ committed:
Added Task class and the ability to mark/unmark tasks

cheahTJ committed:
Added ToDo, Events, Deadlines classes. Also added parser class to deal with user inputs

cheahTJ committed:
Added input and expected output for TextUiTesting

cheahTJ committed:
Added a custom exception class and also handling all sorts of user incorrect/incomplete inputs

cheahTJ committed:
Added the delete feature, and deal with its corresponding possible error from user

cheahTJ committed:
Save and load data from the hard disk

cheahTJ committed:
Make Deadline object take in and output formatted DateTime object

cheahTJ committed:
Introduce more classes to OOP the app

cheahTJ committed:
Organise the classes into suitable packages

cheahTJ committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

cheahTJ committed:
Create JUnit Tests for Parser and TaskList classes, tested with Gradle

cheahTJ committed:
Rename bot to pingmebot, created JAR file

cheahTJ committed:
Added javadoc to all the methods in all the classes

cheahTJ committed:
Fix code according to coding standards

The code base has inconsistency in the coding

The code base needs to be changed to ensure consistency and readability

Modify code base to be according to coding standards

cheahTJ committed:
Add find functionality to the application

The user is unable to find task in his/her list

Implementing this find functionality allows user to find his/her

A find parser and a findMatchingTask() function has been created to understand user's find command and find matching task

cheahTJ committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find functionality to the application

# Conflicts:
#	src/main/java/pingmebot/Pingme.java
#	src/main/java/pingmebot/UI.java

cheahTJ committed:
Resolve merge conflict when merging level-9 branch with master

cheahTJ committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Fix code according to coding standards

# Conflicts:
#	src/main/java/pingmebot/Parser.java
#	src/main/java/pingmebot/PingMe.java
#	src/main/java/pingmebot/PingMeException.java
#	src/main/java/pingmebot/Storage.java
#	src/main/java/pingmebot/TaskList.java
#	src/main/java/pingmebot/UI.java

cheahTJ committed:
Resolve merge conflict with coding standard branch and add new test cases into parser test and task list test

cheahTJ committed:
Create GUI for chatbot

There is no GUI for the chatbot and some logic can be better made to be OOP style.

With a GUI, it is easier for user to key in their commands. With more OOP, the code base is easier to read.

Use JavaFX to create the UI. Create Command and its corresponding subclasses to make the code more OOP.

JavaFX is utilised as it is incorporating well with Gradle. OOP is used to make the code more bearable to read.

Parser and TaskList Test cases were also modified to test the changed code in the respective classes.

chaaaaun commented on README.md:
```diff
@@ -13,7 +13,7 @@ Prerequisites: JDK 11, update Intellij to the most recent version.
    1. If there are any further prompts, accept the defaults.
 1. Configure the project to use **JDK 11** (not other versions) as explained in [here](https://www.jetbrains.com/help/idea/sdk.html#set-up-jdk).<br>
    In the same dialog, set the **Project language level** field to the `SDK default` option.
-3. After that, locate the `src/main/java/Duke.java` file, right-click it, and choose `Run Duke.main()` (if the code editor is showing compile errors, try restarting the IDE). If the setup is correct, you should see something like the below as the output:
```
Thanks for being thorough and updating the readme too :)

chaaaaun commented on src/main/java/pingmebot/PingMeException.java:
```diff
@@ -0,0 +1,17 @@
+/**
+ * Responsible for throwing exception that is specific to the bot application.
+ */
+public class PingMeException extends Exception {
```
Can consider extending PingMeException to encapsulate the different kinds of errors your application runs into, so you have an easier time if you want to handle them differently.

chaaaaun commented on src/main/java/pingmebot/task/Events.java:
```diff
@@ -0,0 +1,68 @@
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from:" + start + " to:" + end + ")";
```
Could try to be more consistent with whether or not to add `this` in front of instance fields.

chaaaaun commented on src/test/java/pingmebot/ParserTest.java:
```diff
@@ -0,0 +1,75 @@
+
+public class ParserTest {
+    @Test
+    public void parseToDoCommandTest() throws PingMeException {
```
Could follow the snake case naming convention so it is easier to see what is the input and expected output for each test.

chaaaaun commented on src/main/java/pingmebot/UI.java:
```diff
@@ -0,0 +1,173 @@
+     * Sends welcoming message to the user.
+     */
+    public void showWelcome() {
+        String greetingMsg = "Hello! I'm PingMeBot\n" + "What can I do for you?";
```
Could extract this to be a constant, perhaps in a separate class or at the top of this class, so that it is easier to find and change if you need to.

chaaaaun reviewed (commented):
Overall really nice job, I especially liked how methodically you've split different parts of the application into different packages and the good use of OOP

jayllo-c commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# pingmebot.PingMe project template
```
I love the refactored changes but note that sometimes the refactoring may not always work properly, like in this example here. How can we avoid unwanted changes in future?

jayllo-c commented on build.gradle:
```diff
@@ -0,0 +1,57 @@
+plugins {
```
Would you like to try installing the checkstyle plugin for Gradle? Here's the [link to the guide](https://se-education.org/guides/tutorials/checkstyle.html).

jayllo-c commented on src/main/java/pingmebot/Parser.java:
```diff
@@ -0,0 +1,241 @@
+package pingmebot;
+
+import pingmebot.command.*;
```
Is importing the whole package in line with the coding standard?

jayllo-c commented on src/main/java/pingmebot/Parser.java:
```diff
@@ -0,0 +1,241 @@
+            }
+
+        } else {
+            throw new PingMeException("I don't understand your command. Try writing: deadline (task description) /by (d/m/yyyy HHmm format)");
```
How many words long is a line acceptably long before we have to wrap it around?

jayllo-c commented on src/main/java/pingmebot/Parser.java:
```diff
@@ -0,0 +1,241 @@
+ */
+public class Parser {
+    protected String userInput;
+    protected ArrayList<String> words;
```
Good use of plural name for this collection

jayllo-c reviewed (commented):
Overall very minor coding standard violations were found, good job for naming variables correctly. The most notable mistake to the coding standard was the long lines found throughout the code. Other than that good job!

cheahTJ committed:
Customise chatbot

The chatbot is looking quite dull

Styling the chatbot makes it look more vibrant

Create CSS file and edit the corresponding fxml files

This is done to edit the appearance of the GUI

cheahTJ committed:
Add checkstyle to enforce coding standards

The code base does not have a code enforcer to ensure coding standard is adhered to

Reason for change is so that cleaner code can be standardised

Add checkstyle plug-ins into build gradle to introduce checkstyle to enforce coding standard

The plug-ins are added to the build gradle file so that gradle can help automate this checking process

cheahTJ committed:
Add Assertions into the code base

The codebase does not have assertions

Assertions need to be introduced to define assumptions about the program state so that the runtime can verify them

Add assert keywords to certain parts of the code

This is to ensure that some variables need to be of a certain state for the program to continue smoothly

cheahTJ committed:
Improve code quality

The current code base is not well written

The code base needs to change to ensure readability in long run

Checkstyle is enforced and SLAP is used

cheahTJ committed:
Merge pull request #2 from cheahTJ/A-CodeQuality

Improve code quality

cheahTJ committed:
Merge branch 'master' into A-Assertions

* master:
  Improve code quality

# Conflicts:
#	src/main/java/pingmebot/Parser.java

cheahTJ committed:
Merge pull request #3 from cheahTJ/A-Assertions

A assertions

cheahTJ committed:
Create postpone feature

User is unable to postpone their task

Flexibility needs to be given to user to postpone tasks

Create new methods in various classes to postpone tasks

These methods create the logic for postponing tasks

cheahTJ committed:
Merge pull request #4 from cheahTJ/BCD-Extension

Create postpone feature

cheahTJ committed:
Improve coding quality

There are ways to improve code quality

It needs to change to make code readable

Use SLAP and abstraction, remove nesting

cheahTJ committed:
Add more testing

New methods added need to be tested

So that the methods be proven to work

Add more testings to ParserTest class

More methods are added to the Parser class

cheahTJ committed:
Add user guide

The code base does not have its own user guide

The guide can teach users to start using the app

Write user guide in README

It is a convenient place for users

cheahTJ committed:
Abstract out lengthier method

Some methods are still quite lengthy

They make it harder to follow the logic

Use SLAP to abstract some longer method

This is to make the method readable

cheahTJ committed:
Amend some GUI text representation

It might not look so presentable

It is to make the GUI reply look more readable

Amend UI class

It is because UI class handles all the UI

cheahTJ committed:
Update README.md

cheahTJ committed:
Update README.md
